### ISSUE TITLE
"401 Unauthorized" with `freenas-api-nfs`, same `apiKey` is fine with `freenas-api-iscsi`

### ISSUE THREAD

**Issue by xeijin** [democratic-csi/democratic-csi#161]
Hello, I am running into an issue where creating new volumes fails with `freenas-api-nfs` even though the same `apiKey` is used to successfully provision volumes with `freenas-api-iscsi`.

I am using Truenas SCALE 22.0.

`Warning  ProvisioningFailed    2m6s (x8 over 4m13s)  org.x8s-democratic-csi.nfs_nfs-democratic-csi-controller-6679868485-kptpv_1b1d7b4d-8cba-4823-b4eb-5fef8279946f  failed to provision volume with StorageClass "truenas-nfs-csi": rpc error: code = Internal desc = Error: "401: Unauthorized"`

```yaml
---
driver: freenas-api-nfs
instance_id: 66a8aa95-faef-46cf-ae78-e9d6f112aea1
httpConnection:
    protocol: https
    host: truenas.netwk.io
    port: 443
    # use only 1 of apiKey or username/password
    # if both are present, apiKey is preferred
    # apiKey is only available starting in TrueNAS-12
    # username: ${SECRET_TRUENAS_API_USER}
    apiKey: ${SECRET_TRUENAS_API_KEY}
    # password:
    # allow self-signed certificates
    allowInsecure: true
    # use apiVersion 2 for TrueNAS-12 and up (will work on 11.x in some scenarios as well)
    # leave unset for auto-detection
    # apiVersion: 2
zfs:
    # can be used to override defaults if necessary
    # the example below is useful for TrueNAS 12
    # cli:
    #  sudoEnabled: true
    #
    #  leave paths unset for auto-detection
    #  paths:
    #    zfs: /usr/local/sbin/zfs
    #    zpool: /usr/local/sbin/zpool
    #    sudo: /usr/local/bin/sudo
    #    chroot: /usr/sbin/chroot
    # can be used to set arbitrary values on the dataset/zvol
    # can use handlebars templates with the parameters from the storage class/CO
    # datasetProperties:
    #  "org.freenas:description": "{{ parameters.[csi.storage.k8s.io/pvc/namespace] }}/{{ parameters.[csi.storage.k8s.io/pvc/name] }}"
    #  "org.freenas:test": "{{ parameters.foo }}"
    #  "org.freenas:test2": "some value"
    datasetParentName: hayward/x8s/nfs/vols
    # do NOT make datasetParentName and detachedSnapshotsDatasetParentName overlap
    # they may be siblings, but neither should be nested in the other
    detachedSnapshotsDatasetParentName: hayward/x8s/nfs/snaps
    datasetEnableQuotas: true
    datasetEnableReservation: false
    datasetPermissionsMode: "0777"
    datasetPermissionsUser: 0
    datasetPermissionsGroup: 0
    #datasetPermissionsAcls:
    #- "-m everyone@:full_set:allow"
    #- "-m u:kube:full_set:allow"
nfs:
    shareCommentTemplate: "{{ parameters.[csi.storage.k8s.io/pvc/namespace] }}-{{ parameters.[csi.storage.k8s.io/pvc/name] }}"
    shareHost: 10.10.100.5
    shareAlldirs: false
    shareAllowedHosts: []
    shareAllowedNetworks: []
    shareMaprootUser: root
    shareMaprootGroup: root
    shareMapallUser: ""
    shareMapallGroup: ""
```

**Comment by travisghansen** [comment]
Welcome! Is that var a literal? Variable expansion doesn’t work in the yaml files like that (currently).

**Comment by xeijin** [comment]
> Welcome! Is that var a literal? Variable expansion doesn’t work in the yaml files like that (currently).

Hey, it does work I'm using a [flux](https://fluxcd.io/) template repo where the secret values are enumerated as environment variables I believe. My `freenas-api-iscsi` config is the same (and works)

**Comment by travisghansen** [comment]
Maybe exec into the controller container and make sure the rendered file has what you expect.

I also would suggest making sure you’re using the same image tag for both deployments to make sure I don’t have a regression in the code. The code for both drivers is the same so it’s unlikely to be related to using nfs vs iscsi.

**Comment by xeijin** [comment]
Hello, sorry for the delay had a chance to try this today.

I noticed that the `driver` value in helm I was using for the controller was set to `democraticcsi/democratic-csi:latest` as the image, so I tried changing this to the `v1.5.4` to match the node driver version, now I get a 401 unathorized for both iSCSI and nfs, and even reverting back to `latest` doesnt bring iSCSI back to life. Strange.

I've managed to exec into the container, but how do I check if the value has rendered? If I do a `kubectl -n democratic-csi get secret truenas-scale-auth-api -oyaml` I can see my API key just fine.
